client/src/send_and_confirm_transactions_in_parallel.rs

```diff
@@ -0,0 +1,469 @@
+
+    let progress_bar = with_spinner.then(|| {
+        let progress_bar = spinner::new_progress_bar();
+        progress_bar.set_message("Setting up...");
+        progress_bar
+    });
+
+    // blockhash anc blockheight update task
+    let block_data_task = create_blockhash_data_updating_task(
+        rpc_client.clone(),
+        blockhash_data_rw.clone(),
+        current_blockheight.clone(),
+    );
+
+    let transaction_map = Arc::new(DashMap::<Signature, TransactionData>::new());
+    let error_map = Arc::new(DashMap::new());
+    let confirmed_transactions = Arc::new(AtomicU32::new(0));
+    // tasks which confirms the transactions that were sent
+    let transaction_confirming_task = create_transaction_confirmation_task(
+        rpc_client.clone(),
+        current_blockheight.clone(),
+        transaction_map.clone(),
+        error_map.clone(),
+        confirmed_transactions.clone(),
+    );
+
+    // transaction sender task
+    let total_transactions = messages.len();
+    let mut initial = true;
+    let signing_count = resign_txs_count.unwrap_or(1);
+    for expired_blockhash_retries in (0..signing_count).rev() {
+        // only send messages which have not been confirmed
+        let messages_with_index: Vec<(usize, Message)> = if initial {
+            initial = false;
+            messages.iter().cloned().enumerate().collect()
+        } else {
+            // remove all the confirmed transactions
+            transaction_map
+                .iter()
+                .map(|x| (x.index, x.message.clone()))
+                .collect()
+        };
+
+        if messages_with_index.is_empty() {
+            break;
+        }
+
+        // clear the map so that we can start resending
+        transaction_map.clear();
+
+        sign_all_messages_and_send(
+            &progress_bar,
+            rpc_client.clone(),
+            &tpu_client,
+            transaction_map.clone(),
+            error_map.clone(),
+            messages_with_index,
+            blockhash_data_rw.clone(),
+            signers,
+            confirmed_transactions.clone(),
+            total_transactions,
+        )
+        .await?;
+
+        // wait till block height till all the transactions are confirmed
+        confirm_transactions_till_block_height_and_resend_unexpired_transaction_over_tpu(
+            &progress_bar,
+            &tpu_client,
+            transaction_map.clone(),
+            current_blockheight.clone(),
+            confirmed_transactions.clone(),
+            total_transactions,
+        )
+        .await;
+
+        if transaction_map.is_empty() {
+            break;
+        }
+
+        if let Some(progress_bar) = &progress_bar {
+            progress_bar.println(format!(
+                "Blockhash expired. {expired_blockhash_retries} retries remaining"
+            ));
+        }
+    }
+
+    block_data_task.abort();
+    transaction_confirming_task.abort();
+    if transaction_map.is_empty() {
```

Comment:
good that you moved this `transaction_map` into a separate loop because having it in code like in https://github.com/solana-labs/solana/blame/7f2f0136bdad8025bf182679d29a792722ab4fd5/tpu-client/src/nonblocking/tpu_client.rs#L543 make it hard to read